<div align='center' >
<p><font size='70'><strong>mlPEA User Manual</strong></font></p>
<font size='100'>(version 1.0)</font>
</div>

- mlPEA is a user-friendly, full-functionality pipeline specifically designed to enhance the processing, analysis, and interpretation of m6A-Seq in non-model plants by leveraging machine learning.
- mlPEA comprises four functional modules: **Data Preprocessing, Transcriptome Construction, m<sup>6</sup>A Calling, and Functional Exploration**.
- mlPEA was powered with an advanced  packaging technology, which enables compatibility and portability.
- mlPEA project is hosted on http://github.com/cma2015/mlPEA
- mlPEA docker image is available at http://hub.docker.com/r/malab/mlpea
- mlPEA server can be accessed via http://mlpea.omstudio.cloud

## m<sup>6</sup>A Calling Module

This module provides step-by-step functions required for epitranscriptome reads mapping and identification of RNA modifications.

#### Align Reads to Genome

Bowtie2 is wrapped to align epitranscriptome reads to genome. 

| **Tools**   | **Description**                                              | **Input**                                                    | **Output**                        | **Time (test data)** | **Reference**             |
| ----------- | ------------------------------------------------------------ | ------------------------------------------------------------ | --------------------------------- | -------------------- | ------------------------- |
| **Bowtie2** | Bowtie2 is a short read aligner which achieves a combination of high speed, sensitivity and accuracy by combining the strengths of the full-text minute index with the flexibility and speed of hardware-accelerated dynamic programming algorithms, therefore bowtie2 is suitable for large genomes | Epitranscriptome sequencing reads in FASTQ format and reference genome sequences in FASTA format | Read alignments in SAM/BAM format | ~5 mins              | (Grabherr *et al*., 2011) |

#### Identify RNA Modifications

**Identify RNA Modifications** implements three pipelines.

| **Tools**                       | **Description**                                              | **Input**                                                    | **Output**                                                   | **Time (test data)** | **Reference**                             |
| ------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | -------------------- | ----------------------------------------- |
| **Peak Calling**                | used the SlidingWindow method with default options. To call m6A peaks, each transfrag was scanned using a 100-bp sliding window with 25-bp steps. | Input the transcripts in FASTA format;Input the ReadsCoverage file in IP sample;Input the ReadsCoverage file in input sample | The enriched peak region matrix in BED format                | ~5 mins              | (Zhai *et al*, 2018)                      |
| **Merge Biological Replicates** | obtain consistent RNA modifications among three biological replicates. | Peak regions for biological Replicates                       | Consistent peak regions among multiple biological replicates | ~10 mins             | in-house scripts                          |
| **High Confidence Peak**        | All peaks derived from the HC transcripts are designated as positive samples and utilized as input for the weakly supervised learning framework to facilitate model training. | Transcripts   in FASTA format;The enriched peak region matrix in BED format | High condfidence Peak region in BED format                   | ~20mins              | (Huang *et al*, 2021; Song *et al*, 2024) |

## Align Reads to Genome

#### Input

- **Epitranscriptome sequencing reads in FASTQ format**
- **Assembled transcripts in FASTA format**


#### Output
- **Alignments in BAM format**
- **Alignment summary generated by bowtie2**
- **ReadsCoverage summary in TXT format generate by BEDtools**

#### How to use this function

- The following screenshot shows us how to use this function

![](../img/3.5.align.png)

## Peak Calling

In this function, Peak calling was performed using PEA (Zhai *et al*, 2018)ï¼Œwhich used the SlidingWindow method with default options. To call m6A peaks, each transfrag was scanned using a 100-bp sliding window with 25-bp steps.

#### Input

- **Input the transfrags file:** Input the transfrags in FASTA format
- **Input the ReadsCoverage file in IP sample:** Input ReadsCoverage file in TXT format
- **Input the ReadsCoverage file in input sample:** Input ReadsCoverage file in TXT format


#### Output

- **The enriched peak region matrix in BED format**

#### How to use this function

- The following screenshot shows us how to use this function

![](../img/3.6.peak_calling.png)

##  **Merge Biological Replicates**

In this function, mlPEA obtained consistent RNA modifications among three biological replicates.

#### Input

- **Peak regions for biological Replicates**


#### Output

- **Consistent peak regions among multiple biological replicates**  

#### How to use this function

- The following screenshot shows us how to use this function

![](../img/3.7.merge_peak.png)

## High Confidence Peak

All peaks derived from the HC transcripts are designated as positive samples and utilized as input for the weakly supervised learning framework to facilitate model training.

#### Input

- **Input the transfrags file:** Input the transfrags in FASTA format
- **Input the peak region file:** The enriched peak region matrix in BED format


#### Output

- **High condfidence Peak region in BED format**

#### How to use this function

- The following screenshot shows us how to use this function

![](../img/3.8.high_confidence.png)
